src/ImageSharp.Web/Providers/PhysicalFileSystemProvider.cs

```diff
@@ -94,13 +87,23 @@ public Task<IImageResolver> GetAsync(HttpContext context)
         /// <returns><see cref="string"/> representing the fully qualified provider root path.</returns>
         internal static string GetProviderRoot(PhysicalFileSystemProviderOptions providerOptions, string webRootPath, string contentRootPath)
         {
-            string providerRoot = string.IsNullOrWhiteSpace(providerOptions.ProviderRootPath)
-                ? webRootPath
-                : providerOptions.ProviderRootPath;
-
-            return Path.IsPathFullyQualified(providerRoot)
+            string providerRoot = providerOptions.ProviderRootPath ?? webRootPath ?? "wwwroot";
+            string fullPath = Path.IsPathFullyQualified(providerRoot)
```

Comment:
Here's where is gets wacky. 
https://github.com/dotnet/aspnetcore/blob/c85baf8db0c72ae8e68643029d514b2e737c9fae/src/Hosting/Hosting/src/Internal/HostingEnvironmentExtensions.cs#L31-L63

If `PhysicalFileSystemProviderOptions.ProviderRootPath` and `WebHostOptions.WebRoot` are both null and `wwwroot` doesn't exist then setting `wwwroot` as a fallback will mean that our provider then points to a folder that doesn't exist and all images requests will silently fail.

I would suggest we throw here if both values are null.

Reply:
In that case the original image can't be found/doesn't exist and nothing will happen... I'd rather see it update the `Match` callback to `false`, so it doesn't even try to use this provider anymore!

Reply:
That still means it silently fails. You want to capture configuration errors at startup with explicit messages. This could be the only provider installed just misconfigured.

Reply:
There's profound difference here. 

We're not the framework which is explicitly designed to work without the static folder as that folder and static files are not a requirement. 

We're the polar opposite. If someone want to add ImageSharp.Web they have to explicitly register the library at startup and that expresses intent. They intend to serve images via the middleware and have registered the physical provider (implicitly or explicitly). That provider needs a location from which to serve the images in order to function. If they have not provided a location and have deleted the wwwroot folder then that becomes a case of misconfiguration because they have expressed intent by deleting the default folder.

Reply:
Fair points, I've updated (and simplified) both the provider and cache implementations to throw an exception when the directory can't be determined (either not explicitly set in the options or the `WebRootPath` is not set, which would indicate the `wwwroot` directory doesn't exist).

Reply:
After refactoring this to the current state, I actually think it indeed makes sense that users either have to explicitly configure the provider/cache root path _or_ ensure the web root path is set (which will be the case when the default wwwroot directory exists or it's also explicitly set).

The only thing that might need to be reconsidered is whether ImageSharp should ensure those root directories exist/are created inside the constructors (doing expensive operations and causing side-effects inside constructors is considered bad practice). This also shouldn't be required, as the provider won't find any files inside the empty directory and the cache will create it before writing files anyway...

Reply:
I would tend to agree but wouldn't say these are particularly expensive operations. They're basically null checks and some string mangling.